### DISCUSSION TITLE
Question about the example mesh1D.py

### DISCUSSION THREAD

**Discussion by CalebDmArcher** [usnistgov/fipy#949] [Q&A] [answered]
I have some questions in [examples/diffusion/mesh1D.py](https://www.ctcms.nist.gov/fipy/examples/diffusion/generated/examples.diffusion.mesh1D.html) about using `sweep()` in the case that 'diffusivity is not only non-uniform'. Hope someone can help me.

The example uses a while loop (`while res > 1e-6:`) to solve the steady-state solution; the example also sets a sweeping number per step and runs for a certain number of steps for the transient-state solution. I was trying to use the while loop in the transient state for each step and turns out it does not work. I noticed that if I use the while loop in the transient state, the solution will always stay at the solution of the first step and will not update anymore. 

To better show my confusion, I wrote a code for comparison in the attachment.

In this code, phi[0] stores the values that are calculated by using the while loop. And phi[1] stores the values that are calculated by setting the sweeping value of 5 per step. I used a total step number of 1000 which is a relatively large number. (I can see that the plot of phi[1] is indeed identical to the plot of the steady-state result in the example which shows the consistency between the steady-state result and transient-state result.)

I am very confused as to why the while loop does not work for the transient-state case. I thought the while loop only makes the results more accurate. Can anyone point out where I got it wrong?

[comparison.zip](https://github.com/usnistgov/fipy/files/12613853/comparison.zip)

**Comment by guyer** [comment, marked as answer]
You need to reset the residual to something big at the beginning of every sweep loop, otherwise the `while` exits without doing anything:
```python
res1 = 1e+10
while res1 > 1e-6:
    res1 = eq_t1.sweep(var=phi[0],
                       dt=timeStepDuration)
```

**Reply by CalebDmArcher** [reply]
Thank you for answering the question. I have a follow-up question: Is it because if we write the code in the wrong way (which is without re-initiate the value of "residual"), the program will not enter the 'while' loop after the first 'for' loop ("residual" will not satisfy the 'while' loop entering condition forever) so that there will be no more sweeping happening and thus the phi values will stay at the first time step?

**Reply by guyer** [reply]
Yes, after the first time through, `res1` will be left at 1e-7 or something. At the next step, it's still 1e-7 and the `while` loop evaluates as `False` and continues to for the next 999 steps.
